modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/ExpressionFactory.java

```diff
@@ -104,22 +113,42 @@
     <T> Supplier<T> execute(RexNode node);
 
     /**
-     * Creates {@link Scalar}, a code-generated expressions evaluator.
+     * Creates {@link SingleScalar}, a code-generated expressions evaluator.
      *
      * @param node Expression.
      * @param type Row type.
-     * @return Scalar.
+     * @return SingleScalar.
      */
-    default Scalar scalar(RexNode node, RelDataType type) {
+    default SingleScalar scalar(RexNode node, RelDataType type) {
         return scalar(ImmutableList.of(node), type);
     }
 
     /**
-     * Creates {@link Scalar}, a code-generated expressions evaluator.
+     * Creates {@link BiScalar}, a code-generated expressions evaluator.
+     *
+     * @param node Expression.
+     * @param type Row type.
+     * @return BiScalar.
+     */
+    default BiScalar biScalar(RexNode node, RelDataType type) {
+        return biScalar(ImmutableList.of(node), type);
+    }
+
+    /**
+     * Creates {@link SingleScalar}, a code-generated expressions evaluator.
+     *
+     * @param nodes Expressions.
+     * @param type Row type.
+     * @return SingleScalar.
+     */
+    SingleScalar scalar(List<RexNode> nodes, RelDataType type);
+
+    /**
+     * Creates {@link BiScalar}, a code-generated expressions evaluator.
      *
      * @param nodes Expressions.
      * @param type Row type.
-     * @return Scalar.
+     * @return BiScalar.
      */
-    Scalar scalar(List<RexNode> nodes, RelDataType type);
+    BiScalar biScalar(List<RexNode> nodes, RelDataType type);
```

Comment:
Looks like this method is redundant and never will be used, it's enough just `biScalar(RexNode node, RelDataType type)` with single node.

Reply:
if i remove it here, i will need to wrap it here .. 
digest(List<RexNode> nodes

Reply:
Yes, you'll need to wrap it in `biScalar` method before pass to `digest` and `compile`, but there will be minus 1 redundant public method (BTW, looks like both `scalar` and `biScalar` methods should not be public at all)

Reply:
catch it ! thanks ! fixed.